README.rst
===============================
Tox Matrix
===============================

.. image:: https://badge.fury.io/py/tox-matrix.png
    :target: http://badge.fury.io/py/tox-matrix

.. image:: https://travis-ci.org/slafs/tox-matrix.png?branch=master
        :target: https://travis-ci.org/slafs/tox-matrix

.. image:: https://pypip.in/d/tox-matrix/badge.png
        :target: https://crate.io/packages/tox-matrix?version=latest


Project for creating tox.ini files with multiple dependencies.

* Free software: BSD license
* Documentation: http://tox-matrix.rtfd.org.

A little utility script to generate ``tox.ini`` files.
It should be handy in case where your project have multiple different dependencies
with different version and you would like to test all available configurations of them.


Quick start
--------------

Install it::

    pip install tox-matrix


Now let's assume you'd like to test your project against different versions
of ``Django`` (latest 1.5.X and 1.6.X ) and ``Fabric`` (latest 1.7.X and 1.8.X)
using ``python2.7`` and ``python3.3``.
You could do something like this::

    tox-matrix generate -d Django -v 1.5,1.6,1.7 -d Fabric -v 1.7,1.8,1.9 -p 2.7 -p 3.3

Which should generate this ``tox.ini`` file content to stdout::

    [tox]
    envlist = py27-A, py27-B, py27-C, py27-D, py33-A, py33-B, py33-C, py33-D

    [testenv]
    commands = py.test

    [testenv:py27-A]
    basepython = python2.7
    deps = {[testenv]deps}
        Django>=1.5,<1.6
        Fabric>=1.7,<1.8

    [testenv:py27-B]
    basepython = python2.7
    deps = {[testenv]deps}
        Django>=1.5,<1.6
        Fabric>=1.8,<1.9

    [testenv:py27-C]
    basepython = python2.7
    deps = {[testenv]deps}
        Django>=1.6,<1.7
        Fabric>=1.7,<1.8

    [testenv:py27-D]
    basepython = python2.7
    deps = {[testenv]deps}
        Django>=1.6,<1.7
        Fabric>=1.8,<1.9

    [testenv:py33-A]
    basepython = python3.3
    deps = {[testenv]deps}
        Django>=1.5,<1.6
        Fabric>=1.7,<1.8

    [testenv:py33-B]
    basepython = python3.3
    deps = {[testenv]deps}
        Django>=1.5,<1.6
        Fabric>=1.8,<1.9

    [testenv:py33-C]
    basepython = python3.3
    deps = {[testenv]deps}
        Django>=1.6,<1.7
        Fabric>=1.7,<1.8

    [testenv:py33-D]
    basepython = python3.3
    deps = {[testenv]deps}
        Django>=1.6,<1.7
        Fabric>=1.8,<1.9


Features
--------

* Generate tox.ini files
* Pin versions either with exact ones or by range
* Use a Jinja2 template for your tox.ini file


.. image:: https://d2weczhvl823v0.cloudfront.net/slafs/tox-matrix/trend.png
   :alt: Bitdeli badge
   :target: https://bitdeli.com/free



HISTORY.rst
.. :changelog:

History
-------

0.1.1 (2014-03-26)
++++++++++++++++++

* Fixes for installation and tests


0.1.0 (2014-03-23)
++++++++++++++++++

* First release on PyPI.


tests/test_cli.py
#!/usr/bin/env python

from toxmatrix.cli import print_default_template
from toxmatrix.template import DEFAULT_TEMPLATE


def test_default(capfd):
    # use a pytest fixture here
    arguments = {}
    print_default_template(arguments)

    resout, reserr = capfd.readouterr()

    assert reserr == ''

    assert resout == DEFAULT_TEMPLATE


tests/test_lib.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_lib
----------------------------------

Tests for `toxmatrix.lib` module.
"""

from toxmatrix.lib import generate_matrix, pin_version


def test_generate_matrix():

    python_versions = ['2.7', '3.3']
    packages = ['Django', 'Fabric']
    version_steps = [['1.5', '1.6', '1.7'], ['1.5', '1.6', '1.7', '1.8']]
    pin_exact = False

    ret = generate_matrix(python_versions, packages, version_steps, pin_exact)

    expected = (
        ('python2.7',
            (
                ('Django>=1.5,<1.6', 'Fabric>=1.5,<1.6'),
                ('Django>=1.5,<1.6', 'Fabric>=1.6,<1.7'),
                ('Django>=1.5,<1.6', 'Fabric>=1.7,<1.8'),
                ('Django>=1.6,<1.7', 'Fabric>=1.5,<1.6'),
                ('Django>=1.6,<1.7', 'Fabric>=1.6,<1.7'),
                ('Django>=1.6,<1.7', 'Fabric>=1.7,<1.8'),
            )
         ),
        ('python3.3',
            (
                ('Django>=1.5,<1.6', 'Fabric>=1.5,<1.6'),
                ('Django>=1.5,<1.6', 'Fabric>=1.6,<1.7'),
                ('Django>=1.5,<1.6', 'Fabric>=1.7,<1.8'),
                ('Django>=1.6,<1.7', 'Fabric>=1.5,<1.6'),
                ('Django>=1.6,<1.7', 'Fabric>=1.6,<1.7'),
                ('Django>=1.6,<1.7', 'Fabric>=1.7,<1.8'),
            )
         )
    )

    assert ret == expected


def test_generate_matrix_exact():

    python_versions = ['2.7', '3.3']
    packages = ['Django', 'Fabric']
    version_steps = [['1.5', '1.6', '1.7'], ['1.5', '1.6', '1.7', '1.8']]
    pin_exact = True

    ret = generate_matrix(python_versions, packages, version_steps, pin_exact)

    expected = (
        ('python2.7',
            (
                ('Django==1.5', 'Fabric==1.5'),
                ('Django==1.5', 'Fabric==1.6'),
                ('Django==1.5', 'Fabric==1.7'),
                ('Django==1.5', 'Fabric==1.8'),
                ('Django==1.6', 'Fabric==1.5'),
                ('Django==1.6', 'Fabric==1.6'),
                ('Django==1.6', 'Fabric==1.7'),
                ('Django==1.6', 'Fabric==1.8'),
                ('Django==1.7', 'Fabric==1.5'),
                ('Django==1.7', 'Fabric==1.6'),
                ('Django==1.7', 'Fabric==1.7'),
                ('Django==1.7', 'Fabric==1.8'),
            )
         ),
        ('python3.3',
            (
                ('Django==1.5', 'Fabric==1.5'),
                ('Django==1.5', 'Fabric==1.6'),
                ('Django==1.5', 'Fabric==1.7'),
                ('Django==1.5', 'Fabric==1.8'),
                ('Django==1.6', 'Fabric==1.5'),
                ('Django==1.6', 'Fabric==1.6'),
                ('Django==1.6', 'Fabric==1.7'),
                ('Django==1.6', 'Fabric==1.8'),
                ('Django==1.7', 'Fabric==1.5'),
                ('Django==1.7', 'Fabric==1.6'),
                ('Django==1.7', 'Fabric==1.7'),
                ('Django==1.7', 'Fabric==1.8'),
            )
         )
    )

    assert ret == expected


def test_pin_package_range():

    package = 'Django'
    version_steps = ['1.5', '1.6', '1.7']
    exact = False

    ret = pin_version(package, version_steps, exact)
    expected = ('Django>=1.5,<1.6', 'Django>=1.6,<1.7')
    assert ret == expected


def test_pin_package_exact():

    package = 'Django'
    version_steps = ['1.5', '1.6', '1.7']
    exact = True

    ret = pin_version(package, version_steps, exact)
    expected = ('Django==1.5', 'Django==1.6', 'Django==1.7')
    assert ret == expected


tests/test_template.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_template
----------------------------------

Tests for `toxmatrix.template` module.
"""

from toxmatrix.template import generate_tox_ini, DEFAULT_TEMPLATE
from jinja2 import Template


def test_default_template():

    matrix = (
        ('python2.7',
            (
                ('Django>=1.5,<1.6', 'Fabric>=1.5,<1.6'),
                ('Django>=1.6,<1.7', 'Fabric>=1.5,<1.6'),
            )
         ),
        ('python3.3',
            (
                ('Django>=1.5,<1.6', 'Fabric>=1.5,<1.6'),
                ('Django>=1.6,<1.7', 'Fabric>=1.5,<1.6'),
            )
         )
    )

    tpl = Template(DEFAULT_TEMPLATE)

    expected = '''[tox]
envlist = py27-A, py27-B, py33-A, py33-B

[testenv]
commands = py.test
deps = pytest

[testenv:py27-A]
basepython = python2.7
deps = {[testenv]deps}
    Django>=1.5,<1.6
    Fabric>=1.5,<1.6

[testenv:py27-B]
basepython = python2.7
deps = {[testenv]deps}
    Django>=1.6,<1.7
    Fabric>=1.5,<1.6

[testenv:py33-A]
basepython = python3.3
deps = {[testenv]deps}
    Django>=1.5,<1.6
    Fabric>=1.5,<1.6

[testenv:py33-B]
basepython = python3.3
deps = {[testenv]deps}
    Django>=1.6,<1.7
    Fabric>=1.5,<1.6
'''

    ret = generate_tox_ini(matrix, tpl)
    assert ret == expected

    # the same thing should be without passing the remplate
    ret2 = generate_tox_ini(matrix)
    assert ret2 == expected


toxmatrix/cli.py
#!/usr/bin/env python
'''
tox-matrix

Usage:
    tox-matrix generate (-d PACKAGE -v VERSION_STRINGS)... [-p PYVERSION]... [options]
    tox-matrix default

Options:
  generate                    generate tox.ini file with given options
  default                     print default template

  -h --help                   Show this screen.
  --version                   Show version.
  -d PACKAGE, --dep=PACKAGE   Package name as a dependency.

  -v VERSION_STRINGS, --ver=VERSION_STRINGS  Comma seperated list of versions for the given package.

  -p PYVERSION, --pyver=PYVERSION  [default: 2.7 3.3] Python versions for basepython option
                                   (can be given multiple times)

  --template=TEMPLATE          A custom template (full path to a file) to generate a tox.ini file contents.
                               Use `default` to print a default template.

  --exact                      [default: False] Pin exact versions of packages.
'''

from __future__ import print_function
import toxmatrix
from toxmatrix.lib import generate_matrix
from toxmatrix.template import generate_tox_ini
from docopt import docopt
from jinja2 import Environment, FileSystemLoader
import sys
import os


def print_default_template(arguments):
    '''
    just print a default template
    '''
    print(toxmatrix.template.DEFAULT_TEMPLATE, end='')
    return 0


def generate_toxini_contents(arguments):
    '''
    prepare arguments and call `generate_matrix` and `generate_tox_ini`
    '''
    packages = arguments.get('--dep', [])
    version_strings = arguments.get('--ver', [])
    pin_exact = arguments.get('--exact', False)
    py_versions = arguments.get('--pyver', [])
    versions = []
    for ver in version_strings:
        versions.append(ver.split(','))
    matrix = generate_matrix(py_versions, packages, versions, pin_exact)

    template = None
    template_filepath = arguments.get('--template', None)
    if template_filepath is not None:
        directory, filename = os.path.split(template_filepath)
        jinjaenv = Environment(loader=FileSystemLoader(directory or '.'))
        template = jinjaenv.get_template(filename)

    res = generate_tox_ini(matrix, template)
    print(res)
    return 0


def main():
    arguments = docopt(__doc__, version=toxmatrix.__version__)

    print_default = arguments.get('default', False)
    generate_toxini = arguments.get('generate', False)
    ret_code = 0

    if print_default:
        ret_code = print_default_template(arguments)

    if generate_toxini:
        ret_code = generate_toxini_contents(arguments)

    sys.exit(ret_code)

if __name__ == '__main__':
    main()


toxmatrix/template.py
#!/usr/bin/env python
'''
Module for creating tox.ini files
'''
from string import ascii_uppercase
from jinja2 import Template

DEFAULT_TEMPLATE = '''[tox]
envlist = {% for toxenv in envs %}{{ toxenv.name }}{% if not loop.last %}, {% endif %}{% endfor %}

[testenv]
commands = py.test
deps = pytest
{% for toxenv in envs %}
[testenv:{{ toxenv.name }}]
basepython = {{ toxenv.basepython }}
deps = {[testenv]deps}{%- for dep in toxenv.deps  %}
    {{ dep }}
{%- endfor %}
{% endfor %}
'''


def generate_tox_ini(matrix, template=None):
    '''
    Generate a tox.ini file contents from a given matrix using a template

    :param matrix: as returned from `toxmatrix.lib.generate_matrix`
    :param template: a template to render tox.ini with Jinja2
    '''

    tox_env = {
        'name': None,
        'basepython': None,
        'deps': [],
    }

    envs = []
    env_suffixes = ascii_uppercase
    if template is None:
        template = Template(DEFAULT_TEMPLATE)

    for basepython, dep_lists in matrix:
        for deps, suffix in zip(dep_lists, env_suffixes):
            env = tox_env.copy()
            env['basepython'] = basepython
            env['name'] = basepython.replace('thon', '').replace('.', '') + '-' + suffix
            env['deps'] = deps
            envs.append(env)

    rendered = template.render(envs=envs)
    return rendered


toxmatrix/lib.py
#!/usr/bin/env python
'''
Module for generic helpers
'''
from itertools import product


def pin_version(package_name, version_steps, exact=False):
    '''
    create a tuple of strings with package and pinned versions.

    :param package_name: package name to be pinned
    :param version_steps: an iterable of versions for the given package name
    :param exact: same as in `generate_matrix`
    '''

    exact_format = '{0}=={1}'
    nonexact_format = '{0}>={1},<{2}'
    ret = []

    if exact:
        for version_string in version_steps:
            ret.append(exact_format.format(package_name, version_string))
    else:
        current = None
        previous = None

        for version_string in version_steps:
            current = version_string
            if previous is not None:
                ret.append(nonexact_format.format(package_name, previous, current))

            previous = current

    return tuple(ret)


def generate_matrix(python_versions, packages, version_steps, pin_exact=False):
    '''
    returns a matrix of packages and its versions including different python
    versions

    :param python_versions: iterable of python version strings e.g. ('2.7', '3.3')

    :param packages: iterable of package names e.g. ('Django', 'Fabric')

    :param version_steps:
        iterable of iterables of version strings from above packages
        e.g. (('1.5', '1.6', '1.7'), ('1.6', '1.7', '1.8'))

    :param pin_exact:
        defines whether to pin exactly above version.

        if *True* then use exact specified versions
        (e.g. to produce something like ('Django==1.5', 'Django==1.6', 'Django==1.7'))

        if *False* then use specified versions as boundaries
        (e.g. to produce something like ('Django>=1.5,<1.6', 'Django>=1.6,<1.7')

    :return list:

        an iterable of two tuples e.g.::

            [('python2.7', (
                ('Django>=1.5,<1.6', 'Fabric>=1.6,<1.7'),
                ('Django>=1.5,<1.6', 'Fabric>=1.7,<1.8'),
                ('Django>=1.6,<1.7', 'Fabric>=1.6,<1.7'),
                ('Django>=1.6,<1.7', 'Fabric>=1.7,<1.8'),
                )
              ),
             ('python3.3', (
                ('Django>=1.5,<1.6', 'Fabric>=1.6,<1.7'),
                ('Django>=1.5,<1.6', 'Fabric>=1.7,<1.8'),
                ('Django>=1.6,<1.7', 'Fabric>=1.6,<1.7'),
                ('Django>=1.6,<1.7', 'Fabric>=1.7,<1.8'),
                )
              ),
            ]

    '''
    ret = []

    dependencies = []
    for package, versions in zip(packages, version_steps):

        package_deps = pin_version(package, versions, pin_exact)
        dependencies.append(package_deps)

    gen = product(*dependencies)
    deps_for_py = tuple(gen)

    for pyver in python_versions:
        semi_deps = tuple(['python{0}'.format(pyver), deps_for_py])
        ret.append(semi_deps)

    return tuple(ret)
